sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -10,14 +10,35 @@
 
 namespace Pulumi
 {
-    internal static class ResourcePackages
+    internal class ResourcePackages
     {
-        private static ImmutableDictionary<string, ImmutableList<(string?, Type)>>? _resourceTypes;
-        private static readonly object _resourceTypesLock = new object();
+        private static Lazy<ImmutableDictionary<string, ImmutableList<(string?, Type)>>> CurrentDomainResourceTypes
+            = new Lazy<ImmutableDictionary<string, ImmutableList<(string?, Type)>>>(
+                () => DiscoverResourceTypes(LoadCurrentDomainReferencedAssemblies()));
+
+        private readonly ImmutableList<Assembly>? _assemblies;
+        private readonly Lazy<ImmutableDictionary<string, ImmutableList<(string?, Type)>>> _resourceTypes;
+
+        /// <param name="assemblies">
+        /// Assemblies that should be used to discover resource types. If <c>null</c>, than <see cref="AppDomain.CurrentDomain"/>
+        /// will be used to find referenced assemblies and discover resource types.
+        /// </param>
+        public ResourcePackages(IEnumerable<Assembly>? assemblies)
```

Comment:
I want to explain this ahead of time.

Our use-case for Pulumi Automation API at my company requires this.. hoping you're OK with it being part of this PR. Essentially our AspNetCore WebAPI will be dynamically loading plugins in order to get our `PulumiFn` equivalent. This means our `PulumiFn` is not part of the same solution as our WebAPI, which means that it is possible for our WebAPI to not have the references that contain those `CustomResource` implementations - as only the plugin loaded at runtime will have those references.

So this allows us to pass through any of the assemblies that the plugin references to get `CustomResource` implementations, without the WebAPI itself having those references, since `AppDomain.CurrentDomain.GetReferencedAssemblies()` would only return the WebAPI's references.